hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/ECUnderReplicationHandler.java

```diff
@@ -0,0 +1,239 @@
+      final Set<ContainerReplica> replicas,
+      final List<ContainerReplicaOp> pendingOps,
+      final ContainerHealthResult result,
+      final int remainingMaintenanceRedundancy) {
+    ContainerInfo container = result.getContainerInfo();
+    ECReplicationConfig repConfig =
+        (ECReplicationConfig) container.getReplicationConfig();
+    final ECContainerReplicaCount replicaCount =
+        new ECContainerReplicaCount(container, replicas, pendingOps,
+            remainingMaintenanceRedundancy);
+
+    ContainerHealthResult currentUnderRepRes = ecContainerHealthCheck
+        .checkHealth(container, replicas, pendingOps,
+            remainingMaintenanceRedundancy);
+
+    LOG.debug("Handling under-replicated EC container: {}", container);
+    if (currentUnderRepRes
+        .getHealthState() != ContainerHealthResult.HealthState
+        .UNDER_REPLICATED) {
+      LOG.info("The container {} state changed and it's not in under"
+              + " replication any more. Current state is: {}",
+          container.getContainerID(), currentUnderRepRes);
+      return emptyMap();
+    }
+    ContainerHealthResult.UnderReplicatedHealthResult containerHealthResult =
+        ((ContainerHealthResult.UnderReplicatedHealthResult)
+            currentUnderRepRes);
+    if (containerHealthResult.isSufficientlyReplicatedAfterPending()) {
+      LOG.info("The container {} with replicas {} is sufficiently replicated",
+          container.getContainerID(), replicaCount.getReplicas());
+      return emptyMap();
+    }
+    if (replicaCount.unRecoverable()) {
+      LOG.warn("The container {} is unrecoverable. The available replicas" +
+          " are: {}.", container.containerID(), replicaCount.getReplicas());
+      return emptyMap();
+    }
+    final ContainerID id = container.containerID();
+    final Map<DatanodeDetails, SCMCommand<?>> commands = new HashMap<>();
+    try {
+     // State is UNDER_REPLICATED
+      final List<DatanodeDetails> deletionInFlight = new ArrayList<>();
+      for (ContainerReplicaOp op : pendingOps) {
+        if (op.getOpType() == ContainerReplicaOp.PendingOpType.DELETE) {
+          deletionInFlight.add(op.getTarget());
+        }
+      }
+      List<Integer> missingIndexes = replicaCount.unavailableIndexes(true);
+      // We got the missing indexes, this is excluded any decommissioning
+      // indexes. Find the good source nodes.
+      if (missingIndexes.size() > 0 && !replicaCount.unRecoverable()) {
+        List<ContainerReplica> sources = replicas.stream().filter(r -> r
+            .getState() == StorageContainerDatanodeProtocolProtos
+            .ContainerReplicaProto.State.CLOSED)
+            // Exclude stale and dead nodes. This is particularly important for
+            // maintenance nodes, as the replicas will remain present in the
+            // container manager, even when they go dead.
+            .filter(r -> ReplicationManager
```

Comment:
the eligible source datanode should satisfy:
1 NodeState is HEALTHY or HEALTHY_READONLY
2 NodeOperationalState is IN_SERVICE

Reply:
Why we need to add  IN_SERVICE? I believe Decom replica can be source too right?

BTW : Here is the LegacyReplicationManager#Underreplication filtering:
```
final List<DatanodeDetails> source = replicas.stream()
    .filter(r ->
        r.getState() == State.QUASI_CLOSED ||
        r.getState() == State.CLOSED)
    // Exclude stale and dead nodes. This is particularly important for
    // maintenance nodes, as the replicas will remain present in the
    // container manager, even when they go dead.
    .filter(r ->
        getNodeStatus(r.getDatanodeDetails()).isHealthy())
    .filter(r -> !deletionInFlight.contains(r.getDatanodeDetails()))
    .sorted((r1, r2) -> r2.getSequenceId().compareTo(r1.getSequenceId()))
    .map(ContainerReplica::getDatanodeDetails)
    .collect(Collectors.toList());
```

Reply:
I am not sure why we did not add HEALTHY_READONLY check. But I just added this check here. If that needs to be added there we can file a separate bug to fix Legacy one too.

Reply:
Seems it was added as part of upgrade. READ_ONLY should probably have been an operational state, not a health state. Sigh.